timesketch/lib/graphs/interface.py

```diff
@@ -0,0 +1,217 @@
+
+        Args:
+            source: (Node) Node to use as source.
+            target: (Node) Node to use as target.
+            label: (str) Label for the node.
+            event: (dict): Elasticsearch event.
+            attributes: (dict) Attributes to add to node.
+        """
+        edge_id_string = ''.join([source.id, target.id, label]).lower()
+        edge_id = hashlib.md5(edge_id_string.encode('utf-8')).hexdigest()
+
+        edge = Edge(source, target, label, attributes)
+
+        if edge.counter < 500:
+            index = event.get('_index')
+            doc_id = event.get('_id')
+            events = edge.attributes.get('events', {})
+            doc_ids = events.get(index, [])
+            doc_ids.append(doc_id)
+            edge.counter += 1
+            events[index] = doc_ids
+            edge.add_attribute('events', events)
+
+        self._edges[edge_id] = edge
+
+    def commit(self):
+        """Commit all nodes and edges to the networkx graph object."""
+        for node_id, node in self._nodes.items():
+            self.nx_instance.add_node(
+                node_id, label=node.label, **node.attributes)
+
+        for _, edge in self._edges.items():
+            label = edge.label + ' ({0:d})'.format(edge.counter)
+            self.nx_instance.add_edge(
+                edge.source.id, edge.target.id, label=label,
+                **edge.attributes)
+
+    def to_cytoscape(self):
+        """Output graph in Cytoscape JSON format.
+
+        Returns:
+            Graph in Cytoscape JSON format.
+        """
+        cy_json = nx.readwrite.json_graph.cytoscape_data(self.nx_instance)
+        return cy_json.get('elements', [])
+
+
+class BaseGraphElement(object):
+    def __init__(self, label='', attributes=None):
+        self.label = label
+        self.attributes = attributes or {}
+        self.id = self.id_from_label(label)
+
+    @staticmethod
+    def id_from_label(label):
+        label = label.lower()
+        return hashlib.md5(label.encode('utf-8')).hexdigest()
+
+    def add_label(self, label):
+        self.label = label
+        self.id = self.id_from_label(label)
+
+    def add_attribute(self, key, value):
+        self.attributes[key] = value
+
+
+class Node(BaseGraphElement):
+    """Graph node object."""
+    # Placeholder until more logic is implemented for Nodes.
+
+
+class Edge(BaseGraphElement):
+    """Graph edge object."""
+    def __init__(self, source, target, label='', attributes=None):
+        self.source = source
+        self.target = target
+        self.counter = 0
+        super(Edge, self).__init__(label, attributes)
+
+
+class BaseGraphPlugin(object):
+    """Base class for a graph."""
+
+    # Name that the graph will be registered as.
+    NAME = 'name'
+
+    # Display name (used in the UI)
+    DISPLAY_NAME = 'display_name'
+
+    # Type of graph. See NetworkX documentation for details:
+    # https://networkx.org/documentation/stable/reference/classes/index.html
```

Comment:
you may also mention that it does not seem you may support every type, just the ones that you defined on top of the file (`GRAPH_TYPES`), or am I not correct?

Reply:
and then there is no actual check to see if the graph type is correct... or valid...

Reply:
good point. Added more text on this, and also a check that will raise if the type is unsupported. This will not happen after development, but nice to have for the developer.